Create Unit class

> [cmutel] I guess a safer way to handle units is also needed - something like a Unit class which handles equality and can spit out a conversion factor on its own. For example, I don't think that 'kg' currently matches 'kilogram'. But please let me put together some data on common conversions first.
> 
> [fjuniorr] The problem that I see handling the conversion with a Unit class would be in the cases when we are dealing with conversions between different dimensions (such as mass to volume) and we need information about the specific flows involved. You can see this for water in [get_conversion_factor](https://github.com/fjuniorr/flowmapper/blob/main/flowmapper/match.py#L51-L57). I know you mentioned that overall these are hard to find so this could be a non-issue.
> 
> [cmutel] Yes. Ideally this would be provided in the glossaries of the different systems... I think we will have maybe 50 of these, but our rules should be designed around the normal unit conversions within dimensions, not the exceptions. You can see the units I have collected through the years [here](https://github.com/brightway-lca/brightway2-io/blob/main/bw2io/units.py). Brightway does unit string normalization because I hate units like "a", and this is another set of flows we will need to match.

I've started to work on this in https://github.com/fjuniorr/flowmapper/pull/51 because it was a good opportunity to create a better way to store original and normalized versions of flow properties (so that any alternative discussed in https://github.com/fjuniorr/flowmapper/issues/46#issuecomment-1843761812 is possible without [hacks](https://github.com/fjuniorr/flowmapper/blob/d520348cd676650255326ae4a13b05d55daebc4b/flowmapper/match.py#L16)).

> something like a Unit class which handles equality and can spit out a conversion factor on its own.

This is working and the tests show the gist of it:

```python
from flowmapper.unit import Unit

def test_init():
    unit = Unit.from_dict(
        {"unitName": {"@xml:lang": "en", "#text": "M2A"}}, "unitName.#text"
    )
    assert unit.value == "square meter-year"
    assert unit.raw_value == "M2A"
    assert unit.raw_object == {"unitName": {"@xml:lang": "en", "#text": "M2A"}}

def test_equals_mass():
    u1 = Unit("kg")
    u2 = Unit("kilogram")

    assert u1 == u2

def test_conversion_factor():
    u1 = Unit("mg")
    u2 = Unit("kg")
    actual = u1.conversion_factor(u2)
    expected = 1e-06
    assert  actual == expected
```

> I think we will have maybe 50 of these, but our rules should be designed around the normal unit conversions within dimensions, not the exceptions.

@cmutel do you think I can remove conversions between dimensions entirely? I left some failing tests until we reach a decision

```bash
FAILED tests/test_get_conversion_factor.py::test_get_conversion_factor_water - assert nan == 0.001
FAILED tests/test_match_mapped_name_differences_with_unit_conversion.py::test_match_non_ionic_state - assert not {'values_changed': {"root[0]['conversion_factor']": {'new_value': 1.2142857142857142, 'old_value': 1.0}}}
```

> You can see the units I have collected through the years [here](https://github.com/brightway-lca/brightway2-io/blob/main/bw2io/units.py). Brightway does unit string normalization because I hate units like "a", and this is another set of flows we will need to match.

I've used [`bw2io.units.normalize_units`](https://github.com/fjuniorr/flowmapper/pull/51/files#diff-1e6b434330ba2562fa6b42911ae7deeaa4aa5df6f10c2f1c214aacac1a2feb59R15) and [`bw2io.units.DEFAULT_UNITS_CONVERSION`](https://github.com/fjuniorr/flowmapper/pull/51/files#diff-1e6b434330ba2562fa6b42911ae7deeaa4aa5df6f10c2f1c214aacac1a2feb59R42) instead of bringing the lists to `flowmapper`.

--- a/flowmapper/flow.py
+++ b/flowmapper/flow.py
@@ -2,6 +2,7 @@ from dataclasses import asdict, dataclass, field
 
 import flowmapper.jsonpath as jp
 
+from .unit import Unit
 from .cas import CAS
 from .context import Context
 from .utils import generate_flow_id
@@ -13,7 +14,7 @@ class Flow:
     name: str = None
     synonyms: list[str] = None
     context: str = None
-    unit: str = None
+    unit: Unit = None
     cas: CAS = None
     fields: dict = field(default_factory=lambda: {"uuid": "uuid", 
                                                   "name": "name", 
@@ -31,7 +32,7 @@ class Flow:
             name = jp.extract(fields['name'], d) if fields.get('name') else None,
             synonyms = jp.extract(fields['synonyms'], d) if fields.get('synonyms') else None,
             context = Context.from_dict(d, fields['context']) if fields.get('context') else None,
-            unit = jp.extract(fields['unit'], d) if fields.get('unit') else None,
+            unit = Unit.from_dict(d, fields['unit']),
             cas = CAS(jp.extract(fields['cas'], d)) if fields.get('cas') else CAS(''),
             fields = fields.copy(),
             raw = d

--- a/flowmapper/flowmap.py
+++ b/flowmapper/flowmap.py
@@ -1,6 +1,7 @@
 from functools import cached_property
 from .flow import Flow
-from .match import match_rules, format_match_result, get_conversion_factor
+from .match import match_rules, format_match_result
+from .unit import Unit
 from tqdm import tqdm
 from typing import Callable
 import pandas as pd
@@ -24,7 +25,7 @@ class Flowmap:
                         result.append(
                             {'from': s,
                              'to': t,
-                             'conversion_factor': get_conversion_factor(s, t),
+                             'conversion_factor': s.unit.conversion_factor(t.unit),
                              'match_rule': rule.__name__,
                              'info': is_match}
                         )
@@ -132,13 +133,13 @@ class Flowmap:
                     'SourceFlowName': map_entry['from'].name,
                     'SourceFlowUUID': source_flow_id,
                     'SourceFlowContext': map_entry['from'].context.full,
-                    'SourceUnit': map_entry['from'].unit,
+                    'SourceUnit': map_entry['from'].unit.raw_value,
                     'MatchCondition': '',
                     'ConversionFactor': map_entry['conversion_factor'],
                     'TargetFlowName': map_entry['to'].name,
                     'TargetFlowUUID': map_entry['to'].uuid,
                     'TargetFlowContext': map_entry['to'].context.full,
-                    'TargetUnit': map_entry['to'].unit,
+                    'TargetUnit': map_entry['to'].unit.raw_value,
                     'MemoMapper': map_entry['info'].get('comment')
                 }
             data.append(row)

--- a/flowmapper/jsonpath.py
+++ b/flowmapper/jsonpath.py
@@ -3,3 +3,14 @@ from glom import Coalesce, glom
 
 def extract(expr, data):
     return glom(data, Coalesce(expr, default = ''))
+
+def root(spec):
+    result = None
+    if isinstance(spec, str):
+        result = spec.split('.')[0]
+    else:
+        try:
+            result = next(iter(spec))
+        except TypeError:
+            pass
+    return result

--- a/flowmapper/match.py
+++ b/flowmapper/match.py
@@ -17,8 +17,8 @@ def format_match_result(s: Flow, t: Flow, conversion_factor: float, match_info: 
     source_result = {
                 s.fields['name']: s.name,
                 source_context_key: s.raw[source_context_key],
-                s.fields['unit']: s.unit
             }
+    source_result.update(s.unit.raw_object)
     if s.uuid:
         source_result.update({s.fields['uuid']: s.uuid})
     
@@ -26,7 +26,7 @@ def format_match_result(s: Flow, t: Flow, conversion_factor: float, match_info: 
                 'uuid': t.uuid,
                 'name': t.name,
                 'context': t.context.full,
-                'unit': t.unit
+                'unit': t.unit.raw_value
             }
     if match_info.get('location'):
         target_result.update({'location': match_info['location']})
@@ -37,36 +37,6 @@ def format_match_result(s: Flow, t: Flow, conversion_factor: float, match_info: 
             'conversion_factor': conversion_factor,
             'comment': match_info['comment']
         }
-    return result
-
-def get_conversion_factor(s:Flow, t: Flow):
-    if s.name == 'Ammonia, as N' and t.name == 'Ammonia':
-        result = 17 / 14
-    elif s.unit == t.unit:
-        result = 1.0
-    elif s.unit == 'm2a' and t.unit == 'm2*year':
-        result = 1.0
-    elif s.unit == 'm3y' and t.unit == 'm3*year':
-        result = 1.0
-    elif s.name.startswith('Water') and t.name.startswith('Water') and s.unit == 'ton' and t.unit == 'm3':
-        result = 1.0
-    elif s.name.startswith('Water') and t.name.startswith('Water') and s.unit == 'g' and t.unit == 'm3':
-        result = 1e-6
-    elif s.name.startswith('Water') and t.name.startswith('Water') and s.unit == 'kg' and t.unit == 'm3':
-        result = 1e-3
-    elif s.name.startswith('Water') and t.name.startswith('Water') and s.unit == 'l' and t.unit == 'm3':
-        result = 1e-3
-    elif s.unit == 'Bq' and t.unit == 'kBq':
-        result = 1e-3
-    elif s.unit == 'ton' and t.unit == 'kg':
-        result = 1e-3
-    elif s.unit == 'g' and t.unit == 'kg':
-        result = 1e3
-    elif s.unit == 'mg' and t.unit == 'kg':
-        result = 1e-6
-    else:
-        result = float('nan')
-
     return result
 
 def match_identical_names_in_synonyms(s: Flow, t: Flow, comment: str = 'Identical synonyms'):

--- a/flowmapper/unit.py
+++ b/flowmapper/unit.py
@@ -1,0 +1,48 @@
+from __future__ import annotations
+from dataclasses import asdict, dataclass, field
+from typing import Any
+import flowmapper.jsonpath as jp
+from bw2io.units import normalize_units, DEFAULT_UNITS_CONVERSION
+
+@dataclass
+class Unit:
+    value: str
+    raw_value: str = ""
+    raw_object: dict = field(default_factory=lambda: {})
+
+    def __post_init__(self):
+        raw_value = self.value
+        self.value = normalize_units(self.value)
+        if raw_value != self.value:
+            self.raw_value = raw_value
+
+    @classmethod
+    def from_dict(cls, d, spec):
+        key = jp.root(spec)
+        value = jp.extract(spec, d)
+        result = Unit(
+            value = normalize_units(value),
+            raw_value = value,
+            raw_object = {key: d.get(key)},
+        )
+        return result
+
+    def to_dict(self):
+        return asdict(self)
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def conversion_factor(self, to: Unit):        
+        if self.value == to.value:
+            result = [1.0]
+        else:
+            result = [
+                multiplier
+                for from_unit, to_unit, multiplier in DEFAULT_UNITS_CONVERSION 
+                if from_unit == self.value and to.value == to_unit
+            ]
+
+        result = float('nan') if not result else result[0]
+        
+        return result

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -32,7 +32,8 @@ dependencies = [
     "glom",
     "tqdm",
     "pandas[excel]",
-    "typer"
+    "typer",
+    "bw2io"
 ]
 
 [project.urls]

--- a/tests/test_flow.py
+++ b/tests/test_flow.py
@@ -36,6 +36,9 @@ def test_flow_with_jsonpath_expr():
     assert flow.context.full == "air/unspecified"
     assert flow.context.primary == "air"
     assert flow.context.secondary == "unspecified"
+    assert flow.unit.value == "kilogram"
+    assert flow.unit.raw_value == "kg"
+    assert flow.unit.raw_object == {'unitName': {'@xml:lang': 'en', '#text': 'kg'}}
 
 
 def test_flow_from_sp_categories():

--- a/tests/test_get_conversion_factor.py
+++ b/tests/test_get_conversion_factor.py
@@ -1,4 +1,3 @@
-from flowmapper.match import get_conversion_factor
 from flowmapper.flow import Flow
 import math
 
@@ -31,7 +30,7 @@ def test_get_conversion_factor(field_mapping):
         field_mapping["target"],
     )
 
-    actual = get_conversion_factor(s, t)
+    actual = s.unit.conversion_factor(t.unit)
     expected = 1e-3
     assert actual == expected
 
@@ -58,7 +57,7 @@ def test_get_conversion_factor_water(field_mapping):
         field_mapping["target"],
     )
 
-    actual = get_conversion_factor(s, t)
+    actual = s.unit.conversion_factor(t.unit)
     expected = 1e-3
     assert actual == expected
 
@@ -88,7 +87,7 @@ def test_get_conversion_factor_m3y(field_mapping):
         field_mapping["target"],
     )
 
-    actual = get_conversion_factor(s, t)
+    actual = s.unit.conversion_factor(t.unit)
     expected = 1
     assert actual == expected
 
@@ -118,7 +117,7 @@ def test_get_conversion_factor_m2a(field_mapping):
         field_mapping["target"],
     )
 
-    actual = get_conversion_factor(s, t)
+    actual = s.unit.conversion_factor(t.unit)
     expected = 1
     assert actual == expected
 
@@ -149,5 +148,5 @@ def test_get_conversion_factor_nan(field_mapping):
         field_mapping["target"],
     )
 
-    actual = get_conversion_factor(s, t)
+    actual = s.unit.conversion_factor(t.unit)
     assert math.isnan(actual)

--- a/tests/test_get_conversion_factor.py
+++ b/tests/test_get_conversion_factor.py
@@ -58,8 +58,7 @@ def test_get_conversion_factor_water(field_mapping):
     )
 
     actual = s.unit.conversion_factor(t.unit)
-    expected = 1e-3
-    assert actual == expected
+    assert math.isnan(actual)
 
 
 def test_get_conversion_factor_m3y(field_mapping):

--- a/tests/test_jsonpath.py
+++ b/tests/test_jsonpath.py
@@ -39,3 +39,10 @@ def test_nested_data_access():
     }
 
     assert jp.extract(("synonym", ["#text"]), data) == ["2-methylbuta-1,3-diene", "methyl bivinyl"]
+
+def test_root():
+    assert jp.root('.unit') == '' # this could be a problem if there are . in column names
+    assert jp.root('unit') == 'unit'
+    assert jp.root('unit.name') == 'unit'
+    assert jp.root('unit.foo.bar') == 'unit'
+    assert jp.root(("synonym", ["#text"])) == 'synonym'

--- a/tests/test_match_biogenic_to_non_fossil.py
+++ b/tests/test_match_biogenic_to_non_fossil.py
@@ -2,9 +2,9 @@ from flowmapper.flow import Flow
 from flowmapper.match import match_biogenic_to_non_fossil
 
 def test_match_biogenic_to_non_fossil():
-    fields = {"name": "name", "context": "context"}
-    s = Flow.from_dict({"name": "Oils, biogenic", "context": "air"}, fields)
-    t = Flow.from_dict({"name": "Oils, non-fossil", "context": "air"}, fields)
+    fields = {"name": "name", "context": "context", "unit": "unit"}
+    s = Flow.from_dict({"name": "Oils, biogenic", "context": "air", "unit": "kg"}, fields)
+    t = Flow.from_dict({"name": "Oils, non-fossil", "context": "air", "unit": "kg"}, fields)
 
     actual = match_biogenic_to_non_fossil(s, t)
     expected = {

--- a/tests/test_match_mapped_name_differences_with_unit_conversion.py
+++ b/tests/test_match_mapped_name_differences_with_unit_conversion.py
@@ -19,7 +19,7 @@ def test_match_non_ionic_state():
         {
             "source": {"name": "Ammonia, as N", "context": "air", "unit": "kg"},
             "target": {"uuid": None, "name": "Ammonia", "context": "air", "unit": "kg"},
-            "conversion_factor": 1.2142857142857142,
+            "conversion_factor": 1.0,
             "comment": "Mapped name differences with unit conversions",
         },
         {

--- a/tests/test_match_names_with_country_codes.py
+++ b/tests/test_match_names_with_country_codes.py
@@ -2,9 +2,9 @@ from flowmapper.flow import Flow
 from flowmapper.match import match_names_with_country_codes
 
 def test_match_names_with_country_codes():
-    fields = {"name": "name", "context": "context"}
-    s = Flow.from_dict({"name": "Ammonia, NL", "context": "air"}, fields)
-    t = Flow.from_dict({"name": "Ammonia", "context": "air"}, fields)
+    fields = {"name": "name", "context": "context", "unit": "unit"}
+    s = Flow.from_dict({"name": "Ammonia, NL", "context": "air", "unit": "kg"}, fields)
+    t = Flow.from_dict({"name": "Ammonia", "context": "air", "unit": "kg"}, fields)
 
     actual = match_names_with_country_codes(s, t)
     expected = {

--- a/tests/test_unit.py
+++ b/tests/test_unit.py
@@ -1,0 +1,35 @@
+from flowmapper.unit import Unit
+import math
+
+def test_init():
+    unit = Unit.from_dict(
+        {"unitName": {"@xml:lang": "en", "#text": "M2A"}}, "unitName.#text"
+    )
+    assert unit.value == "square meter-year"
+    assert unit.raw_value == "M2A"
+    assert unit.raw_object == {"unitName": {"@xml:lang": "en", "#text": "M2A"}}
+
+def test_equals():
+    u1 = Unit("M2A")
+    u2 = Unit("m2*year")
+
+    assert u1 == u2
+
+def test_equals_mass():
+    u1 = Unit("kg")
+    u2 = Unit("kilogram")
+
+    assert u1 == u2
+
+def test_conversion_factor():
+    u1 = Unit("mg")
+    u2 = Unit("kg")
+    actual = u1.conversion_factor(u2)
+    expected = 1e-06
+    assert  actual == expected
+
+def test_nan_conversion_factor():
+    u1 = Unit("bq")
+    u2 = Unit("kg")
+    actual = u1.conversion_factor(u2)
+    assert math.isnan(actual)

